File: scripts/tasks/restore.ts
import * as path from 'path';
import * as fs from 'fs';
import { GetObjectCommand } from '@aws-sdk/client-s3';
import { Readable } from 'stream';
import aws from '../config/aws.js';
import db from '../utils/database.js';
import logger from '../utils/logger.js';
import env from '../config/env.js';

interface RestoreResult {
    success: boolean;
}

class RestoreTask {
    private backupDir: string;

    constructor() {
        this.backupDir = path.join(process.cwd(), 'backups');
    }

    async execute(s3Key: string): Promise<RestoreResult> {
        try {
            // Load environment variables
            env.load();

            if (!s3Key) {
                throw new Error('S3 key is required');
            }

            const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
            const backupPath = path.join(this.backupDir, `restore-${timestamp}.sqlite`);

            // Download from S3
            logger.info('Downloading backup from S3', { s3Key });
            const s3Client = aws.getS3Client();
            const bucket = aws.getBackupBucket();

            const { Body } = await s3Client.send(
                new GetObjectCommand({
                    Bucket: bucket,
                    Key: s3Key,
                }),
            );

            if (!Body) {
                throw new Error('Failed to download backup from S3');
            }

            // Save to local file
            const writeStream = fs.createWriteStream(backupPath);

            // Convert Body to a readable stream if it's not already one
            let readableStream: Readable;
            if (Body instanceof Readable) {
                readableStream = Body;
            } else {
                // Handle different types of Body
                try {
                    // Try to convert to Buffer first
                    const buffer = await Body.transformToByteArray();
                    readableStream = Readable.from(buffer);
                } catch (e) {
                    // Fallback to a simple readable stream
                    readableStream = new Readable({
                        read() {
                            this.push(null);
                        },
                    });
                    logger.warn('Could not convert S3 response to readable stream', {
                        error: e instanceof Error ? e.message : String(e),
                    });
                }
            }

            await new Promise<void>((resolve, reject) => {
                readableStream.pipe(writeStream).on('error', reject).on('finish', resolve);
            });

            // Restore database
            logger.info('Restoring database from backup', { backupPath });
            await db.restore(backupPath);

            // Clean up local backup
            fs.unlinkSync(backupPath);

            logger.info('Restore completed successfully', {
                s3Bucket: bucket,
                s3Key,
            });

            return { success: true };
        } catch (error) {
            logger.error('Restore failed', { error: error instanceof Error ? error.message : String(error) });
            throw error;
        }
    }
}

// If running directly (not imported as a module)
if (require.main === module) {
    const s3Key = process.argv[2];
    if (!s3Key) {
        console.error('Usage: node restore.js <s3-key>');
        process.exit(1);
    }

    new RestoreTask().execute(s3Key).catch((error) => {
        console.error('Restore failed:', error);
        process.exit(1);
    });
}

export default RestoreTask;

File: scripts/tasks/restore.py
import os
import pathlib
import shutil
import zipfile
from datetime import datetime
from typing import Dict, Any

from config.aws import aws
from utils.logger import logger
from config.env import env
from utils.database import db

class RestoreResult:
    def __init__(self, success: bool):
        self.success = success

    def __str__(self) -> str:
        return f"RestoreResult(success={self.success})"

class RestoreTask:
    def __init__(self):
        self.backup_dir = pathlib.Path.cwd() / 'backups'
        self.backup_dir.mkdir(exist_ok=True)

    async def execute(self, s3_key: str = None) -> RestoreResult:
        try:
            # Load environment variables
            env.load()

            if not s3_key:
                raise ValueError('S3 key is required')

            timestamp = datetime.now().isoformat().replace(':', '-').replace('.', '-')
            backup_path = self.backup_dir / f'restore-{timestamp}.sqlite'

            # Download from S3
            logger.info('Downloading backup from S3', s3_key=s3_key)
            s3_client = aws.get_s3_client()
            bucket = aws.get_backup_bucket()

            # Download the file from S3
            s3_client.download_file(bucket, s3_key, str(backup_path))

            # If the file is a zip, extract it
            if str(backup_path).endswith('.zip'):
                temp_dir = self.backup_dir / f'temp_extract_{timestamp}'
                temp_dir.mkdir(exist_ok=True)
                
                with zipfile.ZipFile(backup_path, 'r') as zip_ref:
                    zip_ref.extractall(temp_dir)
                
                # Find the SQLite file in the extracted directory
                sqlite_files = list(temp_dir.glob('*.sqlite'))
                if not sqlite_files:
                    raise FileNotFoundError('No SQLite file found in the backup zip')
                
                # Use the first SQLite file found
                extracted_sqlite = sqlite_files[0]
                
                # Restore database
                logger.info('Restoring database from backup', backup_path=str(extracted_sqlite))
                db.restore(str(extracted_sqlite))
                
                # Clean up
                os.remove(backup_path)
                shutil.rmtree(temp_dir)
            else:
                # Restore database directly
                logger.info('Restoring database from backup', backup_path=str(backup_path))
                db.restore(str(backup_path))
                
                # Clean up local backup
                os.remove(backup_path)

            logger.info('Restore completed successfully', s3_bucket=bucket, s3_key=s3_key)

            return RestoreResult(success=True)
        except Exception as error:
            logger.error('Restore failed', error=str(error))
            raise

if __name__ == '__main__':
    import asyncio
    import sys
    
    if len(sys.argv) < 2:
        print('Usage: python restore.py <s3-key>')
        sys.exit(1)
        
    s3_key = sys.argv[1]
    
    try:
        asyncio.run(RestoreTask().execute(s3_key))
    except Exception as error:
        print('Restore failed:', error)
        sys.exit(1) 